### PR TITLE
1294 categories for organisations

### LINKED ISSUE #1294: Enable categories for organizations + projects?
We got a request from @bakhshir for the following: 

The eScienceCenter PMs would like to add additional information to the projects that show up in the sidebar. Essentially they would like to use the categories we developed for the communities to add top level categories and curate the allowed entries. 

The immediate use case for this is that they want to start recording what infrastructure is used by each project. This is needed for the evaluation report to get an overview of how often we are using SURF infrastructure (and which one), local infrastructure at the partners, national infrastructure such as DAS, public cloud, etc. A second use case is that the PM want to have a curated set of call names, ans label each project with the call(s) that they where funded from. Other use cases are expected.

The best approach for this (I think) it to also enable the categories feature for organizations + projects? That way, each organization maintainer could curate whatever categories they need to classify their software and projects.

### PR DESCRIPTION
**Pull request by dmijatovic**
# Categories for organisations

Closes #1294

Changes proposed in this pull request:
* Organisation maintainer can create custom categories to be used by projects and software of the organisation
* Software and project maintainers can select custom categories specific to organisation      
* Organisation categories are shown on software page in the side panel on the left, under the logo
* Organisation categories are shown on project page in the side panel on the left
* Refactored code to reduce code duplication. ~5% is optimal amount in my opinion, otherwise we will need additional level of abstraction which in my opinion will further increase cognitive load.
* Updated user documentation (software,project and organisation pages)

How to test:
* `make start` to build and generate test data
* navigate to organisations and select first organisation
* login as rsd-admin (first login in dev mode)
* navigate to organisation settings, categories tab
* add categories to organisation:
   * create project specific categories (use "For projects" switch)
   * create software specific categories
   * create shared categories (check both for software and project switch)
* navigate to software item and add organisation that you just added the categories to
* confirm that categories modal appears and you can assign organisation categories 
* select some categories and confirm that these are shown on the software page
* navigate to project item and add organisation that you just added the categories to
* confirm that categories modal appears and you can assign organisation categories 
* select some categories and confirm that these are shown on the project page
* **bonus test:** enable communities in the settings.json and test community categories for software

## Example organisation categories
image

## Example adding organisation categories to software or project
image

## Example organisation categories on project page (bottom of side panel)
image

PR Checklist:

*   [ ] Increase version numbers in `docker-compose.yml`
*   [x] Link to a GitHub issue
*   [x] Update documentation
*   [ ] Tests

### CODE REVIEW

**Comment by jmaassen** [comment]
I seem to have run into a bug. This are the steps I took

- create an empty RSD
- created an organization "Netherlands eScience Center"
- added category "Calls" with subcategories "ASDI2018, ASDI2020, ETEC2020, KD2019"
- added category "Expertise" with subcategory "Computing" and subsubcategories "GPU, Low power, distributed" 
- added to category "Expertise" subcategory "Big data" and subsubcategories "Real time, Heterogeneous" 

Next added project: 

- test
- organization: "Netherlands eScience Center"
- categories: Calls/ASDI2018, Expertise: Computing/GPU and Big Data/Real Time
- publish project

Project page of "test" looks as expected

Next added another project: 

- test2
- organization: "Netherlands eScience Center"
- categories: Calls/ETEC2020, Expertise: Computing/Low Power and Big Data/Heterogeneous
- publish project

Project page of "test2" looks as expected

But the categories of project "test" have now disappeared?

**Comment by dmijatovic** [comment]
@jmaassen Well spotted! I have fixed bug. The PR is (again) ready for testing.

**Review by ewan-escience — COMMENTED** [review]
I will adapt/edit the row level security rules for the tables.

File: frontend/types/Category.ts [hunk 1]
```diff
@@ -1,12 +1,11 @@
 // SPDX-FileCopyrightText: 2023 - 2024 Felix Mühlbauer (GFZ) <felix.muehlbauer@gfz-potsdam.de>
 // SPDX-FileCopyrightText: 2023 - 2024 Helmholtz Centre Potsdam - GFZ German Research Centre for Geosciences
+// SPDX-FileCopyrightText: 2024 Dusan Mijatovic (Netherlands eScience Center)
 // SPDX-FileCopyrightText: 2024 Ewan Cahen (Netherlands eScience Center) <e.cahen@esciencecenter.nl>
 // SPDX-FileCopyrightText: 2024 Netherlands eScience Center
 //
 // SPDX-License-Identifier: Apache-2.0
 
-export type CategoryID = string // NOSONAR ignore: typescript:S6564
```
Comment (ewan-escience) [inline frontend/types/Category.ts:8]: I think the idea of this type is that you can later make the type more specific than it just being a string (like a UUID). Removing it now will make it harder to re-add it later on.

**Review by ewan-escience — CHANGES_REQUESTED** [review]
When adding a new non-root category to an organisation, the two switches to enable/disable software and/or projects for this category is visible, untill you type something in one of the input fields.

Otherwise, nice work!

File: frontend/types/Category.ts (hunk 1 above)
Comment (dmijatovic) [reply frontend/types/Category.ts:8]: Let's do this when the time comes

**Comment by dmijatovic** [comment]
@jmaassen I fixed the modal error Ewan noticed. Can you test again?

**Review by ewan-escience — APPROVED** [review]

**Review by ewan-escience — APPROVED** [review]

**Comment by jmaassen** [comment]
The issue is gone now, and everything works as expected.

**Review by jmaassen — APPROVED** [review]

**Pull request merged**